controllers/logicalvolume_controller.go

```diff
@@ -194,16 +196,58 @@ func (r *LogicalVolumeReconciler) createLV(ctx context.Context, log logr.Logger,
 			return nil
 		}
 
-		resp, err := r.lvService.CreateLV(ctx, &proto.CreateLVRequest{Name: string(lv.UID), DeviceClass: lv.Spec.DeviceClass, SizeGb: uint64(reqBytes >> 30)})
-		if err != nil {
-			code, message := extractFromError(err)
-			log.Error(err, message)
-			lv.Status.Code = code
-			lv.Status.Message = message
-			return err
+		var volume *proto.LogicalVolume
+
+		// Create a snapshot LV
+		if lv.Spec.Source != "" {
+			// accessType should be either "readonly" or "readwrite".
+			if lv.Spec.AccessType != "ro" && lv.Spec.AccessType != "rw" {
+				return fmt.Errorf("invalid access type for source volume: %s", lv.Spec.AccessType)
+			}
+			sourcelv := new(topolvmv1.LogicalVolume)
+			if err := r.Get(ctx, types.NamespacedName{Namespace: lv.Namespace, Name: lv.Spec.Source}, sourcelv); err != nil {
+				if !apierrs.IsNotFound(err) {
+					log.Error(err, "unable to fetch source LogicalVolume")
+					return err
+				}
+				return nil
```

Comment:
Shouldn't this return an error as well?

Reply:
An error should be returned. `createLV` method will be called when the LV doesn't exist. It is odd there aren't any LogicalVolume CR if the source is set.

Reply:
Perhaps this can happen in a race condition - a snapshot created immediately before the source LogicalVolume is deleted.